cpp/cmake/Modules/SetGPUArchs.cmake

```diff
@@ -25,6 +25,20 @@ else()
   list(REMOVE_ITEM SUPPORTED_CUDA_ARCHITECTURES "62" "72")
 endif()
 
+# CMake < 3.20 has a bug in FindCUDAToolkit where it won't
+# properly detect the CUDAToolkit version when
+# find_package(CUDAToolkit) occurs before enable_language(CUDA)
+if(NOT DEFINED CUDAToolkit_VERSION AND CMAKE_CUDA_COMPILER)
+  execute_process(COMMAND ${CMAKE_CUDA_COMPILER} "--version" OUTPUT_VARIABLE NVCC_OUT)
+  if(NVCC_OUT MATCHES [=[ V([0-9]+)\.([0-9]+)\.([0-9]+)]=])
+    set(CUDAToolkit_VERSION_MAJOR "${CMAKE_MATCH_1}")
+    set(CUDAToolkit_VERSION_MINOR "${CMAKE_MATCH_2}")
+    set(CUDAToolkit_VERSION_PATCH "${CMAKE_MATCH_3}")
+    set(CUDAToolkit_VERSION  "${CMAKE_MATCH_1}.${CMAKE_MATCH_2}.${CMAKE_MATCH_3}")
+  endif()
+  unset(NVCC_OUT)
+endif()
+
```

Comment:
FYI, I recently ran into this issue too updating cuML CMake. Did you find any reported CMake issue to track this?

I was able to work around it by ensuring the environment variable `CUDAToolkit_ROOT` is correctly set before calling `cmake` in `./build.sh`. The only benefit of doing it this way is we can add a `--cuda` argument to the build script. Heres the code from `build.sh` if you are interested:
```bash
    CUDA="auto" # Set via `--cuda 11`, `--cuda <PathToCUDA>`
    CUML_CUDA_ROOT=""

    # Now, determine CUDA architecture
    if [[ "${CUDA}" == "auto" ]]; then
        # Need to decide if we are doing anything here
        CUML_CUDA_ROOT="/usr/local/cuda"
    elif [[ "${CUDA}" =~ ^[0-9]{1,2}\.?[0-9]?$ ]]; then
        # Number form. Convert to ##.# form
        major=`echo "$CUDA." | cut -d. -f1`
        minor=`echo "$CUDA." | cut -d. -f2`
        CUML_CUDA_ROOT="/usr/local/cuda-${major}.${minor:-0}"
    else
        CUML_CUDA_ROOT="${CUDA}"
    fi

    # Check for the existance of version.txt which CMake uses to identify CUDA
    if [[ -e "${CUML_CUDA_ROOT}/version.txt" ]]; then
        echo "Building cuML with $(cat "${CUML_CUDA_ROOT}/version.txt")"
        export CUDAToolkit_ROOT="$CUML_CUDA_ROOT"
    elif [[ -x "${CUML_CUDA_ROOT}/bin/nvcc" ]]; then
        echo "Building cuML with CUDA Version $(${CUML_CUDA_ROOT}/bin/nvcc --version | sed -nr 's/^Cuda compilation tools.*V([0-9]{1,2}\.[0-9]\.[0-9]{0,2})$/\1/p')"
        export CUDAToolkit_ROOT="$CUML_CUDA_ROOT"
    else
        echo "Invalid CUDA directory specified: ${CUML_CUDA_ROOT}"
        exit -1
    fi
```

Reply:
@mdemoret-nv 
I corrected the issue upstream https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5787, and it is slated to be part of CMake 3.20

Reply:
Excellent. Funny we both ran into this issue on the same day. Glad a MR is already done.